src/test/app/Credentials_test.cpp

```diff
@@ -0,0 +1,949 @@
+
+                env(credentials::create(subject, issuer, credType2));
+                env.close();
+
+                // Other account can't delete credentials without expiration
+                env(credentials::deleteCred(other, subject, issuer, credType2),
+                    ter(tecNO_PERMISSION));
+                env.close();
+            }
+
+            {
+                testcase("CredentialsDelete fail, time not expired yet.");
+
+                auto jv = credentials::create(subject, issuer, credType);
+                // current time in ripple epoch + 1000s
+                uint32_t const t = env.now().time_since_epoch().count() + 1000;
+                jv[sfExpiration.jsonName] = t;
+                env(jv);
+                env.close();
+
+                // Other account can't delete credentials that not expired
+                env(credentials::deleteCred(other, subject, issuer, credType),
+                    ter(tecNO_PERMISSION));
+                env.close();
+            }
+
+            {
+                testcase("CredentialsDelete fail, no Issuer and Subject.");
+
+                auto jv =
+                    credentials::deleteCred(subject, subject, issuer, credType);
+                jv.removeMember(jss::Subject);
+                jv.removeMember(jss::Issuer);
+                env(jv, ter(temMALFORMED));
+                env.close();
+            }
+
+            {
+                testcase("CredentialsDelete fail, invalid fee.");
+
+                auto jv =
+                    credentials::deleteCred(subject, subject, issuer, credType);
+                jv[jss::Fee] = -1;
+                env(jv, ter(temBAD_FEE));
+                env.close();
+            }
+
+            {
+                testcase("deleteSLE fail, bad SLE.");
+                auto view = std::make_shared<ApplyViewImpl>(
+                    env.current().get(), ApplyFlags::tapNONE);
+                auto ter =
+                    ripple::credentials::deleteSLE(*view, {}, env.journal);
+                BEAST_EXPECT(ter == tecNO_ENTRY);
+            }
+        }
+    }
+
+    void
+    testFeatureFailed(FeatureBitset features)
+    {
+        using namespace test::jtx;
+
+        const char credType[] = "abcde";
+        Account const issuer{"issuer"};
+        Account const subject{"subject"};
+
+        {
+            using namespace jtx;
+            Env env{*this, features};
+
+            env.fund(XRP(5000), subject, issuer);
+            env.close();
+
+            {
+                testcase("Credentials fail, Feature is not enabled.");
+                env(credentials::create(subject, issuer, credType),
+                    ter(temDISABLED));
+                env(credentials::accept(subject, issuer, credType),
+                    ter(temDISABLED));
+                env(credentials::deleteCred(subject, subject, issuer, credType),
+                    ter(temDISABLED));
+            }
+        }
+    }
+
+    void
+    testRPC()
```

Comment:
I would prefer that these tests are in the relevant files for the RPCs - it's a little more straightforward to find them that way, since all other objects have tests done that way.

Reply:
I looked at RPC files, and found them testing something very general with very complicated logic. Such simple test will look weird there